### ISSUE TITLE
Pipe characters passed to Uri.apply not encoded

### ISSUE THREAD

**Issue by emilianogc** [akka/akka-http#1280]
When using the reference configuration, parsing a Uri that contains a '|' in the query portion does not get an escaped result, but when parsing Query directly, the result contains the escaped alternative.

For example:

```
> ${scala_2.12.2}/bin/scala -cp ${jars}/akka-http-core_2.12-10.0.9.jar:\
                                                  ${jars}/akka-actor_2.12-2.5.3.jar:\
                                                  ${jars}/akka-parsing_2.12-10.0.9.jar

Welcome to Scala 2.12.2 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_131).
Type in expressions for evaluation. Or try :help.

scala> import akka.http.scaladsl.model.Uri
import akka.http.scaladsl.model.Uri

scala> import akka.http.scaladsl.model.Uri.Query
import akka.http.scaladsl.model.Uri.Query

scala> val uri = Uri("http://localhost/path?param=A|B")
uri: akka.http.scaladsl.model.Uri = http://localhost/path?param=A|B

scala> val query = Query("param=A|B")
query: akka.http.scaladsl.model.Uri.Query = param=A%7CB

scala> val b = uri.withQuery(uri.query())
b: akka.http.scaladsl.model.Uri = http://localhost/path?param=A%7CB

scala> val c = uri.withQuery(query)
c: akka.http.scaladsl.model.Uri = http://localhost/path?param=A%7CB
```

I'm not sure whether the pipe is a character that requires escaping from the spec, but this was discovered because al least the unescaped form is incompatible with `java.net.URI`.

**Comment by jrudolph** [comment]
Parsing of the query is done lazily so it can paradoxically contain both the decoded and the encoded variant:

```
val uri = Uri("http://localhost/path?param=A|B")
uri: akka.http.scaladsl.model.Uri = http://localhost/path?param=A|B

scala> uri.query()
res1: akka.http.scaladsl.model.Uri.Query = param=A%7CB
```

The current behavior was implemented in https://github.com/akka/akka/pull/18715, but it's not clear to me if the behavior was intended:

 * For most parts of the URI we have canonicalization, i.e. when printed (toString) we'll return a canonicalized version of the original URI, that breaks when the query is not even parsed.
 * afaiu the `rawQueryString` functionality was added to allow supporting the URI spec (RFC 3986) and the stricter specification for queries in HTML / HTTP. We now allow `raw-query-char = VCHAR -- '#'` for any character in the query even in strict parsing mode

I think we should gather a few ideas about we even want for those APIs.

**Comment by jrudolph** [comment]
This has been fixed by a recent change. It will be released as part of akka-http 10.2.0.
